### PR TITLE
options to write R0 and R1 waveforms 

### PR DESCRIPTION
**Pull request by kosack**
For debugging and benchmarking purposes, this adds two options to `DataWriter`: 
* `write_raw_waveforms`: writes the `R0CameraContainer` (waveforms for all channels) to  `/r0/telescope/tel_XXX`
* `write_waveforms`: writes the `R1CameraContainer` (pre-calibrated, gain-selected waveforms) to  `/r1/telescope/tel_XXX`

also add option to transform waveform for more efficient storage

This allows for event-wise benchmarks or cross-checks of data at  all data levels (DL0 currently not supported, as in the current implementation, DL0 and R1 are the same and we have not yet decided how to store sparse waveforms)

image

### CODE REVIEW

**Comment by maxnoe** [comment]
@kosack do you mind if I rebase this branch and fix the pyflakes issue?

**Comment by kosack** [comment]
sure no problem.

**Comment by maxnoe** [comment]
Looks like matplotlib 3.5 dropped

File: ctapipe/io/datawriter.py [hunk 1]
```diff
@@ -128,6 +129,14 @@ class DataWriter(Component):
         help="output filename", default_value=pathlib.Path("events.dl1.h5")
     ).tag(config=True)
 
+    write_raw_waveforms = Bool(
+        help="Store R0 waveforms if available", default_fill_value=False
+    ).tag(config=True)
+
+    write_waveforms = Bool(
+        help="Store R1 waveforms if available", default_fill_value=False
```
[the remark this answers is not in the document]
Comment (kosack) [reply ctapipe/io/datawriter.py:137]: hmm. that may be a completion typo.  It should be default_value.   I'm surprised it didnt' raise an error, though I have not yet written explicit tests.

[the remark this answers is not in the document]
Comment (kosack) [reply ctapipe/io/datawriter.py]: Ah, I see, unknown keywords fail with a warning currently: 
```
DeprecationWarning: metadata {'default_fill_value': 0} was set from the constructor. With traitlets 4.1, metadata should be set using the .tag() method, e.g., Int().tag(key1='value1', key2='value2')
```
So it was not obvious that it was a typo

**Review by maxnoe — CHANGES_REQUESTED** [review]
Shouldn't we just call them `write_r0_waveforms`, `write_r1_waveforms` and `write_dl0_waveforms`? (Adding an option for dl0?)

File: ctapipe/io/datawriter.py [hunk 2]
```diff
@@ -172,6 +181,11 @@ class DataWriter(Component):
 
     overwrite = Bool(help="overwrite output file if it exists").tag(config=True)
 
+    transform_waveform = Bool(default_value=False).tag(config=True)
+    waveform_dtype = Unicode(default_value="int32").tag(config=True)
```
Comment (maxnoe) [inline ctapipe/io/datawriter.py:185]: prefix with `r1` since it only applies to r1?

**Comment by kosack** [comment]
> Shouldn't we just call them write_r0_waveforms, write_r1_waveforms and write_dl0_waveforms? (Adding an option for dl0?)

It was just to avoid jargon since we don't use the data level in the other options either (write_parameters vs write_dl1_parameters), but in the end it may be clearer - I don't mind either way

**Review by maxnoe — APPROVED** [review]

**Review by HealthyPear — APPROVED** [review]

**Comment by maxnoe** [comment]
> It was just to avoid jargon since we don't use the data level in the other options either 

Let me come back to this. I think we really should put the explicit data levels here. These are well defined within the cta data model. Where I did not immediately get to what `waveforms` or `raw_waveforms` refer. Is it `DL0` and `R1`? or `R1` and `R0`?

**Comment by maxnoe** [comment]
But the info is in the help string. So fine, let's merge.

**Pull request merged**